powerauth-java-server/src/main/java/com/wultra/security/powerauth/app/server/service/replay/DefaultReplayVerificationService.java

```diff
@@ -86,32 +87,44 @@ public void checkAndPersistUniqueValue(String protocolVersion, Date requestTimes
      *
      * <p>Version 3.0 and 3.1:
      * <ul>
-     * <li>If TIMESTAMP < CURRENT_TIMESTAMP - EXPIRATION, then reject request</li>
-     * <li>If TIMESTAMP > CURRENT_TIMESTAMP + TIMESTAMP_THRESHOLD, then reject request</li>
+     * <li>If TIMESTAMP < CURRENT_TIMESTAMP - EXTENDED_EXPIRATION, then reject request</li>
+     * <li>If TIMESTAMP > CURRENT_TIMESTAMP + EXTENDED_EXPIRATION, then reject request</li>
      * </ul>
      *
      * <p>Version 3.2+:
      * <ul>
-     * <li>If TIMESTAMP < CURRENT_TIMESTAMP - TIMESTAMP_THRESHOLD, then reject request</li>
-     * <li>If TIMESTAMP > CURRENT_TIMESTAMP + TIMESTAMP_THRESHOLD, then reject request</li>
+     * <li>If TIMESTAMP < CURRENT_TIMESTAMP - EXPIRATION, then reject request</li>
+     * <li>If TIMESTAMP > CURRENT_TIMESTAMP + EXPIRATION, then reject request</li>
      * </ul>
      *
      * @param protocolVersion Protocol version.
      * @param requestTimestamp Request timestamp.
      * @throws GenericServiceException Thrown in case request is rejected due to its timestamp.
      */
-    private void checkTimestamp(String protocolVersion, Date requestTimestamp) throws GenericServiceException {
+    private void checkTimestamp(String protocolVersion, Date requestTimestamp, UniqueValueType uniqueValueType) throws GenericServiceException {
+        if (requestTimestamp == null) {
+            // Rollback is not required, error occurs before writing to database
+            logger.warn("Rejected request due to missing timestamp, protocol version: {}", protocolVersion);
+            throw localizationProvider.buildExceptionForCode(ServiceError.INVALID_REQUEST);
+        }
         final Instant now = Instant.now();
         final Instant requestTime = requestTimestamp.toInstant();
         final Instant limitOldest;
         final Instant limitNewest;
         if ("3.0".equals(protocolVersion) || "3.1".equals(protocolVersion)) {
-            // MAC TOKEN only has extended expiration, ECIES is checked only in protocol versions 3.2+
+            // Only MAC_TOKEN uses extended expiration, ECIES timestamps are checked only in protocol versions 3.2+
+            // The presence or absence of ECIES timestamps is checked before decryption in PowerAuth crypto library:
+            // https://github.com/wultra/powerauth-crypto/blob/develop/powerauth-java-crypto/src/main/java/com/wultra/security/powerauth/crypto/lib/encryptor/ecies/EciesRequestResponseValidator.java
+            if (uniqueValueType != UniqueValueType.MAC_TOKEN) {
```

Comment:
I would not extend the check for irrelevant (theoretically) state, but I have no objections to keeping it.

Reply:
True, this is just to prevent a coding error in the future and enforce what is already written in the JavaDoc.